tools/notification-configuration/notification-creator/NotificationConfigurator.cs

```diff
@@ -152,9 +163,84 @@ public NotificationConfigurator(AzureDevOpsService service, ILogger<Notification
                 }
             }
 
+            if (purpose == TeamPurpose.SynchronizedNotificationTeam)
+            {
+                await SyncTeamWithCodeOwnerFile(pipeline, result, gitHubToAADConverter, persistChanges);
+            }
             return result;
         }
 
+        private async Task SyncTeamWithCodeOwnerFile(BuildDefinition pipeline, WebApiTeam team, GitHubToAADConverter gitHubToAADConverter, bool persistChanges)
+        {
+            using (logger.BeginScope("Team Name = {0}", team.Name))
+            {
+                if (pipeline.Process.Type != PipelineYamlProcessType)
+                {
+                    return;
+                }
+
+                // Get contents of CODEOWNERS
+                logger.LogInformation("Fetching CODEOWNERS file");
+                var managementUrl = new Uri(pipeline.Repository.Properties["manageUrl"]);
+                var codeOwnerEntries = await gitHubService.GetCodeownersFile(managementUrl);
+
+                if (codeOwnerEntries == default)
+                {
+                    logger.LogInformation("CODEOWNERS file not found, skipping sync");
+                    return;
+                }
+
+                var process = pipeline.Process as YamlProcess;
+
+                logger.LogInformation("Searching CODEOWNERS for matching path for {0}", process.YamlFilename);
+                var codeOwnerEntry = CodeOwnersFile.FindOwnersForClosestMatch(codeOwnerEntries, process.YamlFilename);
+                codeOwnerEntry.FilterOutNonUserAliases();
+
+                logger.LogInformation("Matching Contacts Path = {0}, NumContacts = {1}", process.YamlFilename, codeOwnerEntry.Owners.Count);
+
+                // Get set of team members in the CODEOWNERS file
+                var codeownerPrincipals = codeOwnerEntry.Owners
+                    .Select(contact => gitHubToAADConverter.GetUserPrincipalNameFromGithub(contact));
```

Comment:
It looks like we spend a fair amount of time making a lot of calls for resolving these names. We should do what @benbp is doing and pull down all of them at once and then build a cache so we don't have to make so many network calls.

Reply:
That may or may not be faster. We should definitely cache names we've already retrieved, but calling the full list endpoint takes 14+ seconds. It's likely making individual calls are faster, or would be with a cache we build as we go.

Reply:
The opensource API is rather stable compared to devOps. I don't think it caused us problem. I agreed we cache the named if we queried before.

Reply:
Also, we cannot narrow down the list returned for opensource listing API call as what I remembered. I believe query API is most convenient to use currently.

Reply:
@weshaggard we should measure the time in the actual code, but @sima-zhu and I did some profiling in the browser for the same endpoint and it took roughly 100ms per call. I assume a lot of that includes extra data for HTML data vs a JSON body response. So for 100 calls that's 10 seconds, likely even better with API response bodies and caching. That's also if you don't make requests in parallel, which we do. So if we batch 10 concurrent requests it becomes 1 second total.

I worry about adding more and more reliance on the expensive list endpoint and what impact it has on the repos.opensource.microsoft.com site, and whether eventually they'll notice and restrict our usage of it.

Reply:
That sounds reasonable to me I just wanted to make sure that isn't the performance issue here.